cmd/juju/application/refresh.go

```diff
@@ -353,18 +365,36 @@ func (c *refreshCommand) Run(ctx *cmd.Context) error {
 	defer func() { _ = apiRoot.Close() }()
 
 	charmRefreshClient := c.NewCharmRefreshClient(apiRoot)
-	oldURL, oldOrigin, err := charmRefreshClient.GetCharmURLOrigin(ctx, c.ApplicationName)
-	if err != nil {
-		return errors.Trace(err)
-	}
 
 	// There is a timing window where deploy has been called and the charm
 	// is not yet downloaded. Check here to verify the origin has an ID,
-	// otherwise refresh result may be in accurate. We could use the
-	// retry package, but the issue is only seen in automated test due to
-	// speed. Can always use retry if it becomes an issue.
-	if oldOrigin.Source == commoncharm.OriginCharmHub && oldOrigin.ID == "" {
-		return errors.Errorf("%q deploy incomplete, please try refresh again in a little bit.", c.ApplicationName)
+	// otherwise refresh result may be in accurate.
+	// It should only occur in automated test, but if the default retries isn't enough,
+	// the returned error should help to investigate.
+	// Outside automated test, a new refresh will end up to works
+	var (
+		oldURL    *charm.URL
+		oldOrigin commoncharm.Origin
+	)
+
+	err = retry.Call(retry.CallArgs{
+		Func: func() error {
+			oldURL, oldOrigin, err = charmRefreshClient.GetCharmURLOrigin(ctx, c.ApplicationName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if oldOrigin.Source == commoncharm.OriginCharmHub && oldOrigin.ID == "" {
+				return errors.Errorf("%q deploy incomplete, please try refresh again in a little bit.", c.ApplicationName)
+			}
+			return nil
+		},
+		Stop:     ctx.Done(),
+		Attempts: c.RetryGetCharmCount,
+		Delay:    c.RetryGetCharmDelay,
+		Clock:    clock.WallClock,
+	})
+	if err != nil {
+		return errors.Trace(err)
 	}
```

Comment:
We might want to trap the attempt count exceeded and just return the error.

Reply:
Yes, I should have done that, it would avoid me to update tests... Thanks for the nits, I take it.